src/librustc/traits/specialize/specialization_graph.rs

```diff
@@ -291,8 +335,52 @@ impl<'a, 'gcx, 'tcx> Graph {
 
     /// The parent of a given impl, which is the def id of the trait when the
     /// impl is a "specialization root".
-    pub fn parent(&self, child: DefId) -> DefId {
-        *self.parent.get(&child).unwrap()
+    pub fn parent(&self, child: DefId) -> Vec<DefId> {
+        self.parent.get(&child).unwrap().clone()
+    }
+
+    pub fn build_graph(&self) -> (DataStructuresGraph<DefId, String>, DefIdMap<NodeIndex>) {
+        let mut sg_graph: DataStructuresGraph<DefId, String> = DataStructuresGraph::new();
+        let mut nodes_idx = Default::default();
+        for (key, val) in self.parent.iter() {
+            let idx = self.node_idx(&mut sg_graph, &mut nodes_idx, *key);
+            for parent in val {
+                let pidx = self.node_idx(&mut sg_graph, &mut nodes_idx, *parent);
+                sg_graph.add_edge(idx, pidx, format!("fromt {:?} to {:?}", key, parent));
+            }
+        }
+
+        (sg_graph, nodes_idx)
+    }
+
+    /// Return true if impl1 and impl2 are allowed to overlap:
+    /// They have an intersection impl
+    pub fn check_overlap(&self,
+                         sg_graph: &DataStructuresGraph<DefId, String>,
+                         nodes_idx: &DefIdMap<NodeIndex>,
+                         impl1: DefId,
+                         impl2: DefId) -> bool {
+
+        let impl1_idx = *nodes_idx.get(&impl1).unwrap();
+        let impl2_idx = *nodes_idx.get(&impl2).unwrap();
+
+        let impl1_incoming_nodes = sg_graph.nodes_in_postorder(INCOMING, impl1_idx);
+        let impl2_incoming_nodes = sg_graph.nodes_in_postorder(INCOMING, impl2_idx);
```

Comment:
Man, I must be missing something. I don't get why this works. It appears from the code above that the edges go *from* child *to* parent.  

But also, I think that we have to be pretty careful about this check. It's not enough for there to be an impl that specializes both of them: that impl has to cover the full intersection, right? 

For example, consider this:

```rust
trait A { }
trait B { }
trait C { }

trait MyTrait { }

impl<T: A> MyTrait for T { }
impl<T: B> MyTrait for T { }

// This would be OK:
// impl<T: A + B> MyTrait for T { }
// But what about this:
impl<T: A + B + C> MyTrait for T { }
```

Does your code accept that?

Reply:
I'll add your example as a test and let you know.

Reply:
@nikomatsakis My code accept that since A+B+C specializes both A and B but it shouldn't accept it, Right? Because it extends beyond the intersection of A and B. So an additional check is required in order to:
- obtain the intersection between the impls
- check if the impl that specializes both parents fits the intersection